tslearn/early_classification.py

```diff
@@ -0,0 +1,269 @@
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[0]
+        sum_pyhatyck = np.sum(self.pyhatyck_[truncated_t + tau - self.minimum_time_stamp], axis=1)
+        sum_pyhatyck = np.transpose(sum_pyhatyck)
+        sum_global = np.sum((sum_pyhatyck * self.pyck_), axis=0)
+        cost = np.dot(proba_clusters, sum_global)
+        return cost + self._cost_time(truncated_t + tau)
+
+    def cost_function_minimizer(self, end_of_time, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to
+        "end_of_time"
+
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        costs = [self._expected_cost(xt, tau) for tau in range(end_of_time + 1)]
+        tau_star = np.argmin(costs)
+        min_cost = costs[tau_star]
+        return tau_star, min_cost
+
+    def _predict_single_series(self, Xi):
+        """
+        This function classifies a single time series xt
+
+        Parameters
+        ----------
+        xt: vector
+            a time series that is probably incomplete but that nonetheless we want to classify
+        Returns
+        -------
+        int: the class which is predicted
+        int : the time of the prediction
+        float : the probability used for computing the cost
+        float : the loss when classifying
+        """
+
+        time_prediction = self.minimum_time_stamp
+        for t in range(self.minimum_time_stamp, self.__len_X_ + 1):
+            minimum_tau, minimum_loss = self.cost_function_minimizer(
+                end_of_time=self.__len_X_ - t, xt=Xi[:t])
+            if (t == self.__len_X_) or (minimum_tau == t):
+                #Without the transpose here the predict fails miserably
+                Xi = np.transpose(Xi)
+                result = self.classifiers_[t].predict(Xi[:t])
+                result_proba = self.classifiers_[t].predict_proba(Xi[:t])
+                loss_exit = minimum_loss
+                break
+        return result, time_prediction, result_proba, loss_exit
```

Comment:
It seems you do not really need to return `result_proba` and `loss_exit` since they are never used by the calling functions.
In fact, maybe `result_proba` could be kept, since it could be useful if we decide to have a `predict_proba` method at some point, but the last one seems useless.